### PR TITLE
Webdataset performance and cosmetic optimizations

### PR DESCRIPTION
**Pull request by barci2**
## Description
- [ ] **Bug fix** (*non-breaking change which fixes an issue*)
- [ ] **New feature** (*non-breaking change which adds functionality*)
- [ ] **Breaking change** (*fix or feature that would cause existing functionality to not work as expected*)
- [ ] **Refactoring** (*Redesign of existing code that doesn't affect functionality*)
- [X] **Other** (*e.g. Documentation, Tests, Configuration*): Optimization

### What happened in this PR

Added the following optimizations to the code:
 - parallelized data copying in the WDS reader
 - got rid of TarArchive in the WDS loader in favor of direct FileStream
Cosmetic changes:
 - added the repeating components message in WebdatasetLoader::PrepareMetadataImpl

#### Additional information
- Affected modules and functionalities: WebdatasetReader, Executor, Buffer

- Key points relevant for the review: as described in the contents of the PR

### Tests
- [X] Existing tests apply
- [X] New tests added
  - [X] Python tests
  - [ ] GTests
  - [ ] Benchmark
  - [ ] Other
- [ ] N/A

### Documentation
- [X] Existing documentation applies
- [ ] Documentation updated
  - [ ] Docstring
  - [ ] Doxygen
  - [ ] RST
  - [ ] Jupyter
  - [ ] Other
- [ ] N/A

#### Requirements
- [ ] Implements new requirements
- [ ] Affects existing requirements
- [X] N/A

**REQ IDs**: N/A

**JIRA TASK**: DALI-2288

### CODE REVIEW

File: tools/wds2idx.py [hunk 1]
```diff
@@ -43,7 +43,7 @@ class IndexCreator:
     """
 
     tar_block_size = 512
-    index_file_version = "v1.0"
+    index_file_version = "v1.1"
```
[the remark this answers is not in the document]
Comment (barci2) [reply tools/wds2idx.py:46]: doesn't hurt anyone, and it's only a minor change that affects almost nothing (1 test, wds2idx and webdataset_loader.h)

File: dali/operators/reader/webdataset_reader_op.cc [hunk 2]
```diff
@@ -43,14 +44,28 @@ void WebdatasetReader::RunImpl(HostWorkspace& ws) {
...
+    auto& output = ws.OutputRef<CPUBackend>(output_idx);
+    for (int data_idx = 0; data_idx < num_samples; data_idx++) {
+      auto& sample = GetSample(data_idx);
+      ThreadPool::Work copy_task = [output_idx = output_idx, data_idx = data_idx, &output,
+                                    &sample](int) {
+        output[data_idx].SetMeta(sample[output_idx].GetMeta());
+        std::memcpy(output[data_idx].raw_mutable_data(), sample[output_idx].raw_data(),
+                    sample[output_idx].nbytes());
+      };
+      if (threaded) {
```
[the remark this answers is not in the document]
Comment (barci2) [reply dali/operators/reader/webdataset_reader_op.cc:59]: yeah but we will still have to wait for it to finish and lambda transfer is going to be more expensive than just doing that one check (probably?)

File: tools/wds2idx.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (mzient) [reply tools/wds2idx.py:46]: I don't think it hurts anyone - but if anyone (even one of us) has older index files, they'd be painfully bitten by this change.

File: dali/pipeline/data/buffer.h [hunk 3]
```diff
@@ -433,11 +433,11 @@ class DLL_PUBLIC Buffer {
 };
 
 template <typename Backend>
-DLL_PUBLIC double Buffer<Backend>::growth_factor_ = 1.0;
+DLL_PUBLIC double Buffer<Backend>::growth_factor_ = 1.1;
```
Comment (JanuszL) [inline dali/pipeline/data/buffer.h:436]: I have second thought about this change. I think it should go to a separate PR.

Comment (mzient) [reply dali/pipeline/data/buffer.h:436]: Also, this change calls for all tests (L1, L2, L3).

File: tools/wds2idx.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (barci2) [reply tools/wds2idx.py:46]: yeah but I mean that the contents of the index file changed - otherwise I wouldn't be changing the version

File: dali/pipeline/data/buffer.h (hunk 3 above)
Comment (barci2) [reply dali/pipeline/data/buffer.h:436]: alright

File: dali/pipeline/executor/executor.cc [hunk 4]
```diff
@@ -318,11 +318,9 @@ void Executor<WorkspacePolicy, QueuePolicy>::RunHelper(OpNode &op_node, Workspac
     for (int i = 0; i < ws.NumOutput(); i++) {
       auto &desc = output_desc[i];
       if (ws.template OutputIsType<CPUBackend>(i)) {
-        ws.template OutputRef<CPUBackend>(i).Resize(desc.shape);
-        ws.template OutputRef<CPUBackend>(i).set_type(desc.type);
+        ws.template OutputRef<CPUBackend>(i).Resize(desc.shape, desc.type);
```
[the remark this answers is not in the document]
Comment (barci2) [reply dali/pipeline/executor/executor.cc:321]: alright

File: dali/operators/reader/loader/webdataset_loader.cc [hunk 5]
```diff
@@ -157,8 +157,8 @@ WebdatasetLoader::WebdatasetLoader(const OpSpec& spec)
     }
   }
 
-  dtypes_ = spec.HasArgument("dtypes") ? spec.GetRepeatedArgument<DALIDataType>("dtypes")
-                                       : std::vector<DALIDataType>(ext_.size(), DALI_UINT8);
+  dtypes_ = spec.HasArgument("dtypes") ? spec.GetRepeatedArgument<DALIDataType>("dtypes") :
+                                         std::vector<DALIDataType>(ext_.size(), DALI_UINT8);
```
Comment (mzient) [inline dali/operators/reader/loader/webdataset_loader.cc:161]: Why? I find it much less readable this way.

File: dali/operators/reader/loader/webdataset_loader.cc [hunk 6]
```diff
@@ -328,6 +317,12 @@ void WebdatasetLoader::PrepareMetadataImpl() {
             output_indicies_.push_back(output);
             component.outputs.num++;
             was_output_set[output] = true;
+          } else {
+            std::call_once(multiple_files_single_component, [&]() {
+              DALI_WARN(make_string("Multiple components matching the same output for output nr ",
```
Comment (mzient) [inline dali/operators/reader/loader/webdataset_loader.cc:322]:
```suggestion
              DALI_WARN(make_string("Multiple components matching the same output for output # ",
```

File: dali/operators/reader/loader/webdataset_loader.cc [hunk 7]
```diff
@@ -328,6 +317,12 @@ void WebdatasetLoader::PrepareMetadataImpl() {
             output_indicies_.push_back(output);
             component.outputs.num++;
             was_output_set[output] = true;
+          } else {
+            std::call_once(multiple_files_single_component, [&]() {
```
Comment (mzient) [inline dali/operators/reader/loader/webdataset_loader.cc:321]: This will print it only once per WDS reader, not once per occurrence of the problem (or even per output). As a user attempting to fix that, I'd be deeply frustrated to fix the errors one by one. On the other hand, if the user is OK with that, it would be nice to have a way to limit or even silence these warnings.

File: dali/operators/reader/loader/webdataset_loader.cc (hunk 5 above)
Comment (barci2) [reply dali/operators/reader/loader/webdataset_loader.cc:161]: auto formatter, is it an issue?

File: dali/operators/reader/loader/webdataset_loader.cc (hunk 6 above)
Comment (barci2) [reply dali/operators/reader/loader/webdataset_loader.cc:322]: is it really that obvious that a # is an abbreviation for nr?

File: dali/operators/reader/loader/webdataset_loader.cc (hunk 7 above)
Comment (barci2) [reply dali/operators/reader/loader/webdataset_loader.cc:321]: yeah but I don't think it's really a problem, it's just part of the requirements (and if it is a problem then it's really not that hard to deduce what other occurrences are, you just look at the composition of the tar file which you have to rebuild anyways if you wanna get rid of a file in the middle)

File: dali/operators/reader/loader/webdataset_loader.cc (hunk 5 above)
Comment (mzient) [reply dali/operators/reader/loader/webdataset_loader.cc:161]: Well, In general we try not to apply formatting to otherwise unmodified code unless there is an express intention to reformat a piece of code.
Here, the autoformatter destroyed the original intention: aligning the parts of the ternary operator, clearly delineating which part of the expression goes to which case. The dangling colon at the end of the line doesn't read nearly as good.

File: dali/operators/reader/loader/webdataset_loader.cc (hunk 7 above)
Comment (mzient) [reply dali/operators/reader/loader/webdataset_loader.cc:321]: OK.

File: dali/operators/reader/loader/webdataset_loader.cc (hunk 6 above)
Comment (mzient) [reply dali/operators/reader/loader/webdataset_loader.cc:322]: Then use the proper English abbreviation, which is `no.` or skp it altogether and rephrase:
```suggestion
              DALI_WARN(make_string("Multiple components matching output ",
```

This will yield a message:
> Multiple components matching output 0 at line 123 file whatever.idx

File: dali/operators/reader/loader/webdataset_loader.cc [hunk 8]
```diff
@@ -328,6 +317,12 @@ void WebdatasetLoader::PrepareMetadataImpl() {
...
+              DALI_WARN(make_string("Multiple components matching the same output for output nr ",
+                                    output, " at line ", sample.line_number, " file ",
+                                    index_paths_[wds_shard_index]));
```
Comment (mzient) [inline dali/operators/reader/loader/webdataset_loader.cc:324]: I think the file name should be quoted - most, if not all, modern OSes allow spaces in paths.
```suggestion
                                    output, " at line ", sample.line_number, " file \"",
                                    index_paths_[wds_shard_index], "\"."));
```

File: dali/operators/reader/loader/webdataset_loader.cc (hunk 6 above)
Comment (barci2) [reply dali/operators/reader/loader/webdataset_loader.cc:322]: alright

File: dali/operators/reader/loader/webdataset_loader.cc (hunk 8 above)
Comment (barci2) [reply dali/operators/reader/loader/webdataset_loader.cc:324]: alright

File: dali/operators/reader/loader/webdataset_loader.cc (hunk 5 above)
Comment (barci2) [reply dali/operators/reader/loader/webdataset_loader.cc:161]: alright

**Review by mzient — APPROVED** [review]

**Comment by barci2** [comment]
!build
